src/meta/src/controller/fragment.rs

```diff
@@ -1292,22 +1292,47 @@ impl CatalogController {
         Ok(actors)
     }
 
-    /// Get the actor ids, and each actor's upstream actor ids of the fragment with `fragment_id` with `Running` status.
-    pub async fn get_running_actors_and_upstream_of_fragment(
+    /// Get the actor ids, and each actor's upstream source actor ids of the fragment with `fragment_id` with `Running` status.
+    pub async fn get_running_actors_for_source_backfill(
         &self,
         fragment_id: FragmentId,
-    ) -> MetaResult<Vec<(ActorId, ActorUpstreamActors)>> {
+    ) -> MetaResult<Vec<(ActorId, ActorId)>> {
         let inner = self.inner.read().await;
+        let txn = inner.db.begin().await?;
+        let fragment = Fragment::find_by_id(fragment_id)
```

Comment:
Can we then still use `fragment.upstream_fragment_id` but check the `fragment_type_mask` to obtain the desired one?

Reply:
It will rely on a new assumption that the upstream fragments can only have 1 Source fragment. It seems safer, but I'd prefer not to rely on it.